### PR TITLE
feat: red packet confirm dialog

### LINKED ISSUE #3368: [Demand] Red packet confirm dialog
## Catalogue

- [x] UI/UX
- [ ] Mask Core Feature
- [ ] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background
https://www.figma.com/file/gVkQ67y285b4FXVV1KPThN/Twitter?node-id=4312%3A216588

Open confirm dialog before send the red packet.

image

### PR DESCRIPTION
**Pull request by lelenei**
closes #3368

### CODE REVIEW

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/RedPacketForm.tsx [hunk 1]
```diff
@@ -127,7 +125,11 @@ export function RedPacketForm(props: RedPacketFormProps) {
     )
 
     // amount
-    const [rawAmount, setRawAmount] = useState('')
+    const [rawAmount, setRawAmount] = useState(
+        origin?.isRandom
+            ? formatBalance(new BigNumber(origin?.total || 0), origin.token?.decimals ?? 0)
```
Comment (septs) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/RedPacketForm.tsx:130]:
```suggestion
            ? formatBalance(origin?.total ?? 0, origin.token?.decimals ?? 0)
```

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/RedPacketForm.tsx [hunk 2]
```diff
@@ -127,7 +125,11 @@ export function RedPacketForm(props: RedPacketFormProps) {
...
+            : formatBalance(new BigNumber(origin?.total ?? '0').div(origin?.shares || 1), origin?.token?.decimals ?? 0),
```
Comment (septs) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/RedPacketForm.tsx:131]:
```suggestion
            : formatBalance(new BigNumber(origin?.total ?? '0').div(origin?.shares ?? 1), origin?.token?.decimals ?? 0),
```

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                    <Grid item xs={6}>
+                        <Typography variant="body1" color="textSecondary">
+                            {t('plugin_red_packet_amount_per_share')}
+                        </Typography>
+                    </Grid>
+                    <Grid item xs={6}>
+                        <Typography variant="body1" color="textPrimary" align="right">
+                            {`${
+                                settings?.isRandom
+                                    ? formatBalance(new BigNumber(settings?.total || 0), settings?.token?.decimals ?? 0)
```
Comment (septs) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:97]:
```suggestion
                                    ? formatBalance(settings?.total ?? 0, settings?.token?.decimals ?? 0)
```

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                                    : formatBalance(
+                                          new BigNumber(settings?.total || 0).div(settings?.shares || 1),
```
Comment (septs) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:99]:
```suggestion
                                          new BigNumber(settings?.total ?? 0).div(settings?.shares || 1),
```

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+            <Grid item xs={6}>
+                <ActionButton variant="contained" size="large" fullWidth onClick={onBack}>
+                    {t('plugin_red_packet_back')}
+                </ActionButton>
+            </Grid>
+            <Grid item xs={6}>
+                <ActionButton variant="contained" size="large" fullWidth onClick={onCreate}>
+                    {t('plugin_red_packet_send_symbol', {
+                        amount: formatBalance(new BigNumber(settings?.total || 0), settings?.token?.decimals ?? 0),
```
Comment (septs) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:139]:
```suggestion
                        amount: formatBalance(settings?.total ?? 0, settings?.token?.decimals ?? 0),
```

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/RedPacketForm.tsx (hunk 2 above)
Comment (UncleBill) [reply packages/maskbook/src/plugins/RedPacket/SNSAdaptor/RedPacketForm.tsx]: `origin?.total` might be an empty string, then `||` makes sense.

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx (hunk 3 above)
Comment (UncleBill) [reply packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx]: ditto.

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                                    ? formatBalance(new BigNumber(settings?.total ?? 0), settings?.token?.decimals ?? 0)
+                                    : formatBalance(
+                                          new BigNumber(settings?.total ?? 0).div(settings?.shares ?? 1),
+                                          settings?.token?.decimals ?? 18,
+                                      )
+                            } ${settings?.token?.symbol}`}
+                        </Typography>
+                    </Grid>
+                </>
+            )}
```
Comment (UncleBill) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:106]: There are two times of checking of `settings?.isRandom`, the last one is unnecessary, right?

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                    <Grid item xs={6}>
+                        <Typography variant="body1" color="textPrimary" align="right">
+                            {`${
+                                settings?.isRandom
+                                    ? formatBalance(new BigNumber(settings?.total ?? 0), settings?.token?.decimals ?? 0)
+                                    : formatBalance(
+                                          new BigNumber(settings?.total ?? 0).div(settings?.shares ?? 1),
+                                          settings?.token?.decimals ?? 18,
+                                      )
+                            } ${settings?.token?.symbol}`}
```
Comment (septs) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:102]: why not use `<FormattedBalance>`

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/RedPacketForm.tsx [hunk 8]
```diff
@@ -127,9 +125,13 @@ export function RedPacketForm(props: RedPacketFormProps) {
     )
 
     // amount
-    const [rawAmount, setRawAmount] = useState('')
-    const amount = new BigNumber(rawAmount || '0').multipliedBy(pow10(token?.decimals ?? 0))
-    const totalAmount = isRandom ? new BigNumber(amount) : new BigNumber(amount).multipliedBy(shares || '0')
+    const [rawAmount, setRawAmount] = useState(
+        origin?.isRandom
+            ? formatBalance(new BigNumber(origin?.total ?? 0), origin.token?.decimals ?? 0)
```
Comment (septs) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/RedPacketForm.tsx:130]:
```suggestion
            ? formatBalance(origin?.total ?? 0, origin.token?.decimals ?? 0)
```

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx [hunk 9]
```diff
@@ -92,14 +91,11 @@ export function ConfirmRedPacketForm(props: ConfirmRedPacketFormProps) {
...
-                                    : formatBalance(
-                                          new BigNumber(settings?.total ?? 0).div(settings?.shares ?? 1),
-                                          settings?.token?.decimals ?? 18,
-                                      )
-                            } ${settings?.token?.symbol}`}
+                            <FormattedBalance
+                                value={settings?.total}
+                                decimals={settings?.token?.decimals}
+                                symbol={settings?.token?.symbol}
+                            />
```
Comment (septs) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:98]: CleanShot 2021-07-09 at 15 08 39@2x
?????

Comment (septs) [reply packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:99]: @lelenei you unhandling decimalz

**Review by septs — APPROVED** [review]

File: packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                    <Grid item xs={6}>
+                        <Typography variant="body1" color="textSecondary">
+                            {t('plugin_red_packet_amount_per_share')}
+                        </Typography>
+                    </Grid>
+                    <Grid item xs={6}>
+                        <Typography variant="body1" color="textPrimary" align="right">
+                            <FormattedBalance
+                                value={new BigNumber(settings?.total ?? 0).div(settings?.shares ?? 1)}
+                                decimals={settings?.token?.decimals!}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:97]:
```suggestion
                                decimals={settings?.token?.decimals}
```

Comment (guanbinrui) [reply packages/maskbook/src/plugins/RedPacket/SNSAdaptor/confirmRedPacketForm.tsx:97]: refactor: decimals is optional

**Review by septs — APPROVED** [review]

**Comment by zhouhanseng** [comment]
Please wait me to add some works on it.

**Comment by zhouhanseng** [comment]
image

image

image

**Comment by wenluomask** [comment]
Checked.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**